# Create redshift summary `zall` value added catalog: Healpix Version
Stéphanie Juneau (NOIRLab)

## Imports

In [1]:
import os
import sys
#sys.path.insert(0, os.path.join(os.environ['HOME'], 'desihub', 'desispec', 'py'))
import numpy as np

from astropy.table import Table, vstack
from astropy.io import fits

from desispec import coaddition

## HEALPIX coadd redshift catalog: `zall-pix` 

In [2]:
# List all combinations for fuji
surveys_all = ['cmx', 'special',  'sv1',  'sv1',  'sv1',  'sv1',  'sv2',  'sv2',  'sv2',  'sv3',  'sv3',  'sv3']
programs_all = ['other', 'dark', 'backup', 'bright', 'dark', 'other', 'backup', 'bright', 'dark', 'backup', 'bright', 'dark']

N = len(surveys_all)

In [3]:
zall_orig_file = "/global/cfs/cdirs/desi/spectro/redux/fuji/zcatalog/zall-pix-fuji.fits"
zvac_path = "/global/cfs/cdirs/desi/science/gqp/vac/edr/zcat/tests/"
zvac_file = zvac_path+"zall-pix-edr-vac.fits"

## For `zpix` loop over {survey}-{program}

In [20]:
%%time
# re-run 3 (sv1-bright)
for i in np.arange(3,4):

    survey = surveys_all[i]
    program = programs_all[i]

    # Read file with two extensions per each survey-program
    zfile = f"/global/cfs/cdirs/desi/spectro/redux/fuji/zcatalog/zpix-{survey}-{program}.fits"
   
    print(zfile)    
    zcat_single = Table.read(zfile, "ZCATALOG")
    fibermap_input = Table.read(zfile, "EXP_FIBERMAP")
    
    # Keep only rows with TARGETID>0
    zcat_single = zcat_single[zcat_single['TARGETID']>0]
    fibermap_input = fibermap_input[fibermap_input['TARGETID']>0]
    
    #- Sanity check
    Nuniq_id = len(np.unique(zcat_single['TARGETID']))
    
    if Nuniq_id!=len(zcat_single):
        msg = f'ERROR: Not all input TARGETIDs are unique: found {Nuniq_id} unique out of {len(zcat_single)}'
        u, indices, counts = np.unique(zcat_single['TARGETID'], return_index=True, return_counts=True)
        dup = zcat_single['TARGETID'][indices][counts>1]
        print(f"Number of cases with N>1 TARGETIDs: {len(dup)}")
        print(f"with TARGETID<0: {len(dup[dup<0])}")
        print("=========================================")

    # Free memory
    del zcat_single
        
    # Obtain coadded fibermap and exp_fibermap (again)
    tfmap, _ = coaddition.coadd_fibermap(fibermap_input)
    
    tfmap['SURVEY']=survey
    tfmap['PROGRAM']=program
    
    patch_file = f"tmp_patch/zpix-{survey}-{program}-coaddpatch.fits"
    tfmap.write(patch_file, overwrite=True)

/global/cfs/cdirs/desi/spectro/redux/fuji/zcatalog/zpix-sv1-bright.fits
Problematic TARGETID with no good_coadds&good_coords: 39627731599624825
CPU times: user 44min 56s, sys: 5.7 s, total: 45min 2s
Wall time: 45min 7s


### In /global/cfs/cdirs/desi/spectro/redux/fuji/zcatalog/zpix-sv1-bright.fits
Problematic TARGETID with no good_coadds&good_coords: 39627731599624825
`tid = 39627731599624825`

### In /global/cfs/cdirs/desi/spectro/redux/fuji/zcatalog/zpix-sv1-other.fits
```
#Problematic TARGETID with no good_coadds&good_coords: 616094057155789098
#Problematic TARGETID with no good_coadds&good_coords: 39633332794819324
#Problematic TARGETID with no good_coadds&good_coords: 39633336196401313
#Problematic TARGETID with no good_coadds&good_coords: 234520867504138
```

### In /global/cfs/cdirs/desi/spectro/redux/fuji/zcatalog/zpix-sv3-dark.fits
```
Problematic TARGETID with no good_coadds&good_coords: 616088557668271184
Problematic TARGETID with no good_coadds&good_coords: 616088557668271075
Problematic TARGETID with no good_coadds&good_coords: 39627805356461773
Problematic TARGETID with no good_coadds&good_coords: 616088575779275866
Problematic TARGETID with no good_coadds&good_coords: 39627841586859614
Problematic TARGETID with no good_coadds&good_coords: 39627817465416000
Problematic TARGETID with no good_coadds&good_coords: 39627835563837424
Problematic TARGETID with no good_coadds&good_coords: 39627835568032521
Problematic TARGETID with no good_coadds&good_coords: 39627835572227827
Problematic TARGETID with no good_coadds&good_coords: 39627841612025987
Problematic TARGETID with no good_coadds&good_coords: 39627835572229137
Problematic TARGETID with no good_coadds&good_coords: 39627835572228083
Problematic TARGETID with no good_coadds&good_coords: 616088587875648333
Problematic TARGETID with no good_coadds&good_coords: 616088587879842451
Problematic TARGETID with no good_coadds&good_coords: 616088587875648413
Problematic TARGETID with no good_coadds&good_coords: 616088593915445263
Problematic TARGETID with no good_coadds&good_coords: 39627835572228170
Problematic TARGETID with no good_coadds&good_coords: 39627841612026522
Problematic TARGETID with no good_coadds&good_coords: 39627835568033883
Problematic TARGETID with no good_coadds&good_coords: 39627841607831781
Problematic TARGETID with no good_coadds&good_coords: 39627835568033903
Problematic TARGETID with no good_coadds&good_coords: 39627835568034833
Problematic TARGETID with no good_coadds&good_coords: 39627835568033573
Problematic TARGETID with no good_coadds&good_coords: 39627829536621185
Problematic TARGETID with no good_coadds&good_coords: 616088587879842243
Problematic TARGETID with no good_coadds&good_coords: 616088587875647570
Problematic TARGETID with no good_coadds&good_coords: 616088587875647773
Problematic TARGETID with no good_coadds&good_coords: 616088587871454167
Problematic TARGETID with no good_coadds&good_coords: 616088587871453919
Problematic TARGETID with no good_coadds&good_coords: 39627835572225561
Problematic TARGETID with no good_coadds&good_coords: 1083355953102852
Problematic TARGETID with no good_coadds&good_coords: 616088587875647714
Problematic TARGETID with no good_coadds&good_coords: 616088587875648112
Problematic TARGETID with no good_coadds&good_coords: 616088581835850387
Problematic TARGETID with no good_coadds&good_coords: 39627847668598082
Problematic TARGETID with no good_coadds&good_coords: 39627847668598671
Problematic TARGETID with no good_coadds&good_coords: 39627841612026957
Problematic TARGETID with no good_coadds&good_coords: 39627841616217178
Problematic TARGETID with no good_coadds&good_coords: 616088599942661164
Problematic TARGETID with no good_coadds&good_coords: 39627853683229651
Problematic TARGETID with no good_coadds&good_coords: 616088618066248363
Problematic TARGETID with no good_coadds&good_coords: 39627853691619597
Problematic TARGETID with no good_coadds&good_coords: 39627847651819585
Problematic TARGETID with no good_coadds&good_coords: 39627853708395797
Problematic TARGETID with no good_coadds&good_coords: 39627871806820161
Problematic TARGETID with no good_coadds&good_coords: 616088605965681487
Problematic TARGETID with no good_coadds&good_coords: 616088612009673465
Problematic TARGETID with no good_coadds&good_coords: 616088612013868190
Problematic TARGETID with no good_coadds&good_coords: 39627865746051214
Problematic TARGETID with no good_coadds&good_coords: 39627865754441085
```

## Original zall file

In [8]:
#hdul = fits.open(zall_orig_file)
#hdr = hdul[1].header

In [9]:
fits.info(zall_orig_file)

Filename: /global/cfs/cdirs/desi/spectro/redux/fuji/zcatalog/zall-pix-fuji.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU       4   ()      
  1  ZCATALOG      1 BinTableHDU    333   2847435R x 130C   [K, 7A, 6A, J, J, D, D, K, D, 10D, K, 6A, 20A, K, D, J, D, D, E, E, E, K, B, 3A, D, J, I, 8A, J, J, 4A, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, I, E, E, E, E, K, 2A, E, E, E, E, 1A, K, K, K, K, K, K, K, K, K, K, K, K, K, K, K, K, K, K, K, D, D, I, E, I, I, E, E, E, E, D, E, D, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, E, J, L, K, L]   


In [10]:
%%time
# Read input catalog and only keep cases with TARGETID>0
tz = Table.read(zall_orig_file)
tz = tz[tz['TARGETID']>0]

CPU times: user 8.13 s, sys: 1.16 s, total: 9.29 s
Wall time: 9.63 s


## Read in patch files

In [14]:
t_patch = Table()

# N
for i in range(N):
    survey = surveys_all[i]
    program = programs_all[i]

    # Read file per each survey-program
    filename = f"tmp_patch/zpix-{survey}-{program}-coaddpatch.fits"
    t = Table.read(filename)
    
    t_patch = vstack([t_patch, t])

In [15]:
print(len(t_patch))
t_patch[:5]

2451325


<Table length=5>
      TARGETID          SUBPRIORITY     COADD_FIBERSTATUS ... SURVEY PROGRAM
       int64              float64             int32       ... bytes7  bytes6
------------------- ------------------- ----------------- ... ------ -------
 616089230757593610  0.8476382465505894                 0 ...    cmx   other
 616089230757593262  0.9301407809760713                 0 ...    cmx   other
  39628478454170137  0.3927336764704654                 0 ...    cmx   other
 616089230757593291  0.9073811024291156                 0 ...    cmx   other
1152921504619432614 0.14599465692329916                 0 ...    cmx   other

## Replace and add columns based on patch files

In [17]:
%%time
# Check that lenghts and row order are consistent
Nz = len(tz)
Np = len(t_patch)

# Columns exist but values are updated
cols_to_replace = ['MEAN_DELTA_X','RMS_DELTA_X','MEAN_DELTA_Y','RMS_DELTA_Y','MEAN_FIBER_RA', \
                   'STD_FIBER_RA', 'MEAN_FIBER_DEC', 'STD_FIBER_DEC','MEAN_PSF_TO_FIBER_SPECFLUX']

# New columns
cols_to_add = ['MIN_MJD', 'MEAN_MJD', 'MAX_MJD', 'FIRSTNIGHT', 'LASTNIGHT']

if Nz==Np:
    tz.sort(['SURVEY','PROGRAM','TARGETID'])
    t_patch.sort(['SURVEY','PROGRAM','TARGETID'])
    
    # Replace columns listed above
    for col in cols_to_replace:
        tz[col] = t_patch[col]

    # Add new columns listed above
    for col in cols_to_add:
        tz[col] = t_patch[col]

## Adding targeting column patch

In [ ]:
patchfile = zvac_path+"zall-pix-targeting-patch-edr.fits"
tg = Table.read(patchfile)

In [ ]:
tg = tg[tg['SURVEY']!='cmx']  # N=272 excluding CMX
print(f"Check if N=272: {(len(tg)==272)}")

In [1]:
target_cols = ['CMX_TARGET', 'DESI_TARGET', 'BGS_TARGET', 'MWS_TARGET', \
               'SV1_DESI_TARGET', 'SV1_BGS_TARGET', 'SV1_MWS_TARGET', \
               'SV2_DESI_TARGET', 'SV2_BGS_TARGET', 'SV2_MWS_TARGET', \
               'SV3_DESI_TARGET', 'SV3_BGS_TARGET', 'SV3_MWS_TARGET', \
               'SCND_TARGET', 'SV1_SCND_TARGET', 'SV2_SCND_TARGET', 'SV3_SCND_TARGET']

In [ ]:
%%time
for i,tid in enumerate(tg['TARGETID']):
    ii = (tz['TARGETID']==tid)&(tz['SURVEY']==tg['SURVEY'][i])&(tz['PROGRAM']==tg['PROGRAM'][i])
    if len(zpix_before[ii])>1: print(f"ERROR: more than one match!! for {tid}")
    for col in target_cols:
        tz[col][ii] = tg[col][i]

In [ ]:
# Fix DESI_TARGET to be zero for CMX
is_cmx = tz['SURVEY']=='cmx'
tz['DESI_TARGET'][is_cmx] = 0

## Save output file

In [18]:
tz.write(zvac_file)